---
title: "Ruegg et. al., WIFL Genoscape, Data and Code Repository"
author: "Kristen C. Ruegg, Eric C. Anderson, Marius, Mary W, RBay, Eben, and Tom"
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 3
  github_document:
    toc: true
    toc_depth: 3
bibliography: references.bib
---

**Last Updated:** `r Sys.Date()`


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# use this for the the README.md
GPbase <- "https://eriqande.github.io/ruegg-et-al-wifl-genoscape/"

# use this for making the README.html that we will hard link to docs/index.html
Dbase <- "./"

# choose the one that you want here depending on if you are on the
# GitHub README or are making an index.html for the docs directory
hdir <- Dbase

make_link <- function(Rmd, base = hdir) {
  nb <- stringr::str_replace(Rmd, "Rmd$", "html")
  sprintf("%s\n\n(_Compiled RMarkdown HTML document on GitHub Pages:_ [%s](%s%s))\n\n", Rmd, nb, base, nb)
}
```


This repository includes code, data, and some intermediate results to
reproduce the results in Ruegg et al. ("Seasonal niche breadth predicts
population declines in a long-distance migratory bird").
You can get the whole thing by cloning or downloading the repo from
[https://github.com/eriqande/ruegg-et-al-wifl-genoscape](https://github.com/eriqande/ruegg-et-al-wifl-genoscape).

If you are viewing this as the README of a GitHub repository, note that
you can read it in a somewhat friendlier format (i.e., with a table
of contents, etc.) on GitHub pages at:
[https://eriqande.github.io/ruegg-et-al-wifl-genoscape/](https://eriqande.github.io/ruegg-et-al-wifl-genoscape/)


For the whole-genome resequencing data, the production of BAMs and VCFs from the original FASTQ files follows a
standard bwa/GATK pipeline as described in the supplemental methods.  Here we focus on analyses
of the products of that work.  The various steps are described in a series of RMarkdown documents with a
fair bit of explanation/description of the procedures.  Analyses requiring the use
of cluster computing resources are described and documented, but will not be run automatically
by evaluating the RMarkdown document.  Implementing them yourself will
require modification for the vagaries of your own high performance computing system and
preferred mode of parallelization, etc.  Running some of the RMarkdown documents requires
a proper shell.  These were initially developed and run on a Mac. They have also
been tested and confirmed to run on a server running CentOs 7.

Rmarkdown documents and scripts in the 000 series were developed primarily by Eric C. Anderson.



All the RMarkdown documents or scripts can be evaluated en masse by
sourcing the R script `render-numbered-Rmds.R`.  On a fairly old mac laptop the run times
for each are as follows.

Rmarkdown document  | Run time
------------------- | ---------
001-align-coho-genome.Rmd | 1.7 secs
002-allele-frequencies.Rmd (after running initial results on cluster) | 3.3 mins
003-extract-johnson-creek-variants.Rmd | 8.7 mins



# Preliminaries and Dependencies

Below is a description of the needed dependencies.  A script to install all of
them on one of our test clusters is at `000-00-prepare-dependencies.sh`, but it will
likely need some tweaking on your system.



## Geospatial data

Make a directory in the top level of the repository called
`geo-spatial` and then put some things in there:


3. Download some things from Natural Earth Data:
    - Natural Earth II with Shaded Relief, Water, and Drainages raster [https://www.naturalearthdata.com/http//www.naturalearthdata.com/download/10m/raster/NE2_HR_LC_SR_W_DR.zip](https://www.naturalearthdata.com/http//www.naturalearthdata.com/download/10m/raster/NE2_HR_LC_SR_W_DR.zip).  Put or symlink the resulting directory, `NE2_HR_LC_SR_W_DR` into `geo-spatial`.
    - 10m-cultural-vectors, Admin 1 â€“ States, Provinces, Download boundary lines: [https://www.naturalearthdata.com/http//www.naturalearthdata.com/download/10m/cultural/ne_10m_admin_1_states_provinces_lines.zip](https://www.naturalearthdata.com/http//www.naturalearthdata.com/download/10m/cultural/ne_10m_admin_1_states_provinces_lines.zip). Put or symlink the resulting directory, `ne_10m_admin_1_states_provinces_lines` into `geo-spatial`.
    - Finally, get the coastlines: [https://www.naturalearthdata.com/http//www.naturalearthdata.com/download/10m/physical/ne_10m_coastline.zip](https://www.naturalearthdata.com/http//www.naturalearthdata.com/download/10m/physical/ne_10m_coastline.zip), and put the resulting folder, `ne_10m_coastline` into `geo-spatial`.

## Unix Programs

The following programs must be installed and available in the PATH. Versions
used on Eric's laptop appear in parentheses.



## Java jars

The following Java-based programs must be downloaded, and the paths to their
associated Jar files must be listed appropriately in the file `script/java-jar-paths.R`:


## R Packages

Packages must be downloaded from CRAN, BioConductor, and GitHub.  The following
code, which is in `R/install_packages_etc.R`, will download and install the necessary
packages:

```r
`r paste(readLines("R/install_packages_etc.R"), collapse = "\n")`
```

# RMarkdown Documents

The following RMarkdown documents should be evaluated in order.
The script `render-numbered-Rmds.R` will do that (except for `002-allele-frequencies.Rmd`)
when run in the top level
of this repository.
Some RMarkdown documents rely on
outputs from previous ones.  Some of these RMarkdown documents include
calls to Unix utilities, so might not run on non-Unix or non-Linux architectures.

Outputs (figures, tables, R data objects, etc) from each RMarkdown document
are written to the `outputs/XXX` directories.  Intermediate files
written during evaluation of each RMarkdown document are written
to the `intermediates/XXX` directories. To facilitate working between
the cluster and a desktop/laptop, some outputs are written to
the `stored_results/XXX` directories which are version controlled and
included in this repo.


Thumbnails of the figures and tables generated by each RMarkdown document appear
below.  Additionally, at the top of each section is a link to the
compiled (HTML-version) of the RMarkdown document on GitHub Pages.


## `r make_link("001-align-coho-genome.Rmd")`

The products of this analysis go into creation of some of the trees (`009`) and also
in the designation of the alleles as ancestral or not in some of the haplotype rasters (`006`).



## References

