<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1">

<style type="text/css">
@font-face {
font-family: octicons-link;
src: url(data:font/woff;charset=utf-8;base64,d09GRgABAAAAAAZwABAAAAAACFQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABEU0lHAAAGaAAAAAgAAAAIAAAAAUdTVUIAAAZcAAAACgAAAAoAAQAAT1MvMgAAAyQAAABJAAAAYFYEU3RjbWFwAAADcAAAAEUAAACAAJThvmN2dCAAAATkAAAABAAAAAQAAAAAZnBnbQAAA7gAAACyAAABCUM+8IhnYXNwAAAGTAAAABAAAAAQABoAI2dseWYAAAFsAAABPAAAAZwcEq9taGVhZAAAAsgAAAA0AAAANgh4a91oaGVhAAADCAAAABoAAAAkCA8DRGhtdHgAAAL8AAAADAAAAAwGAACfbG9jYQAAAsAAAAAIAAAACABiATBtYXhwAAACqAAAABgAAAAgAA8ASm5hbWUAAAToAAABQgAAAlXu73sOcG9zdAAABiwAAAAeAAAAME3QpOBwcmVwAAAEbAAAAHYAAAB/aFGpk3jaTY6xa8JAGMW/O62BDi0tJLYQincXEypYIiGJjSgHniQ6umTsUEyLm5BV6NDBP8Tpts6F0v+k/0an2i+itHDw3v2+9+DBKTzsJNnWJNTgHEy4BgG3EMI9DCEDOGEXzDADU5hBKMIgNPZqoD3SilVaXZCER3/I7AtxEJLtzzuZfI+VVkprxTlXShWKb3TBecG11rwoNlmmn1P2WYcJczl32etSpKnziC7lQyWe1smVPy/Lt7Kc+0vWY/gAgIIEqAN9we0pwKXreiMasxvabDQMM4riO+qxM2ogwDGOZTXxwxDiycQIcoYFBLj5K3EIaSctAq2kTYiw+ymhce7vwM9jSqO8JyVd5RH9gyTt2+J/yUmYlIR0s04n6+7Vm1ozezUeLEaUjhaDSuXHwVRgvLJn1tQ7xiuVv/ocTRF42mNgZGBgYGbwZOBiAAFGJBIMAAizAFoAAABiAGIAznjaY2BkYGAA4in8zwXi+W2+MjCzMIDApSwvXzC97Z4Ig8N/BxYGZgcgl52BCSQKAA3jCV8CAABfAAAAAAQAAEB42mNgZGBg4f3vACQZQABIMjKgAmYAKEgBXgAAeNpjYGY6wTiBgZWBg2kmUxoDA4MPhGZMYzBi1AHygVLYQUCaawqDA4PChxhmh/8ODDEsvAwHgMKMIDnGL0x7gJQCAwMAJd4MFwAAAHjaY2BgYGaA4DAGRgYQkAHyGMF8NgYrIM3JIAGVYYDT+AEjAwuDFpBmA9KMDEwMCh9i/v8H8sH0/4dQc1iAmAkALaUKLgAAAHjaTY9LDsIgEIbtgqHUPpDi3gPoBVyRTmTddOmqTXThEXqrob2gQ1FjwpDvfwCBdmdXC5AVKFu3e5MfNFJ29KTQT48Ob9/lqYwOGZxeUelN2U2R6+cArgtCJpauW7UQBqnFkUsjAY/kOU1cP+DAgvxwn1chZDwUbd6CFimGXwzwF6tPbFIcjEl+vvmM/byA48e6tWrKArm4ZJlCbdsrxksL1AwWn/yBSJKpYbq8AXaaTb8AAHja28jAwOC00ZrBeQNDQOWO//sdBBgYGRiYWYAEELEwMTE4uzo5Zzo5b2BxdnFOcALxNjA6b2ByTswC8jYwg0VlNuoCTWAMqNzMzsoK1rEhNqByEyerg5PMJlYuVueETKcd/89uBpnpvIEVomeHLoMsAAe1Id4AAAAAAAB42oWQT07CQBTGv0JBhagk7HQzKxca2sJCE1hDt4QF+9JOS0nbaaYDCQfwCJ7Au3AHj+LO13FMmm6cl7785vven0kBjHCBhfpYuNa5Ph1c0e2Xu3jEvWG7UdPDLZ4N92nOm+EBXuAbHmIMSRMs+4aUEd4Nd3CHD8NdvOLTsA2GL8M9PODbcL+hD7C1xoaHeLJSEao0FEW14ckxC+TU8TxvsY6X0eLPmRhry2WVioLpkrbp84LLQPGI7c6sOiUzpWIWS5GzlSgUzzLBSikOPFTOXqly7rqx0Z1Q5BAIoZBSFihQYQOOBEdkCOgXTOHA07HAGjGWiIjaPZNW13/+lm6S9FT7rLHFJ6fQbkATOG1j2OFMucKJJsxIVfQORl+9Jyda6Sl1dUYhSCm1dyClfoeDve4qMYdLEbfqHf3O/AdDumsjAAB42mNgYoAAZQYjBmyAGYQZmdhL8zLdDEydARfoAqIAAAABAAMABwAKABMAB///AA8AAQAAAAAAAAAAAAAAAAABAAAAAA==) format('woff');
}
body {
-webkit-text-size-adjust: 100%;
text-size-adjust: 100%;
color: #333;
font-family: "Helvetica Neue", Helvetica, "Segoe UI", Arial, freesans, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
font-size: 16px;
line-height: 1.6;
word-wrap: break-word;
}
a {
background-color: transparent;
}
a:active,
a:hover {
outline: 0;
}
strong {
font-weight: bold;
}
h1 {
font-size: 2em;
margin: 0.67em 0;
}
img {
border: 0;
}
hr {
box-sizing: content-box;
height: 0;
}
pre {
overflow: auto;
}
code,
kbd,
pre {
font-family: monospace, monospace;
font-size: 1em;
}
input {
color: inherit;
font: inherit;
margin: 0;
}
html input[disabled] {
cursor: default;
}
input {
line-height: normal;
}
input[type="checkbox"] {
box-sizing: border-box;
padding: 0;
}
table {
border-collapse: collapse;
border-spacing: 0;
}
td,
th {
padding: 0;
}
* {
box-sizing: border-box;
}
input {
font: 13px / 1.4 Helvetica, arial, nimbussansl, liberationsans, freesans, clean, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
}
a {
color: #4078c0;
text-decoration: none;
}
a:hover,
a:active {
text-decoration: underline;
}
hr {
height: 0;
margin: 15px 0;
overflow: hidden;
background: transparent;
border: 0;
border-bottom: 1px solid #ddd;
}
hr:before {
display: table;
content: "";
}
hr:after {
display: table;
clear: both;
content: "";
}
h1,
h2,
h3,
h4,
h5,
h6 {
margin-top: 15px;
margin-bottom: 15px;
line-height: 1.1;
}
h1 {
font-size: 30px;
}
h2 {
font-size: 21px;
}
h3 {
font-size: 16px;
}
h4 {
font-size: 14px;
}
h5 {
font-size: 12px;
}
h6 {
font-size: 11px;
}
blockquote {
margin: 0;
}
ul,
ol {
padding: 0;
margin-top: 0;
margin-bottom: 0;
}
ol ol,
ul ol {
list-style-type: lower-roman;
}
ul ul ol,
ul ol ol,
ol ul ol,
ol ol ol {
list-style-type: lower-alpha;
}
dd {
margin-left: 0;
}
code {
font-family: Consolas, "Liberation Mono", Menlo, Courier, monospace;
font-size: 12px;
}
pre {
margin-top: 0;
margin-bottom: 0;
font: 12px Consolas, "Liberation Mono", Menlo, Courier, monospace;
}
.select::-ms-expand {
opacity: 0;
}
.octicon {
font: normal normal normal 16px/1 octicons-link;
display: inline-block;
text-decoration: none;
text-rendering: auto;
-webkit-font-smoothing: antialiased;
-moz-osx-font-smoothing: grayscale;
-webkit-user-select: none;
-moz-user-select: none;
-ms-user-select: none;
user-select: none;
}
.octicon-link:before {
content: '\f05c';
}
.markdown-body:before {
display: table;
content: "";
}
.markdown-body:after {
display: table;
clear: both;
content: "";
}
.markdown-body>*:first-child {
margin-top: 0 !important;
}
.markdown-body>*:last-child {
margin-bottom: 0 !important;
}
a:not([href]) {
color: inherit;
text-decoration: none;
}
.anchor {
display: inline-block;
padding-right: 2px;
margin-left: -18px;
}
.anchor:focus {
outline: none;
}
h1,
h2,
h3,
h4,
h5,
h6 {
margin-top: 1em;
margin-bottom: 16px;
font-weight: bold;
line-height: 1.4;
}
h1 .octicon-link,
h2 .octicon-link,
h3 .octicon-link,
h4 .octicon-link,
h5 .octicon-link,
h6 .octicon-link {
color: #000;
vertical-align: middle;
visibility: hidden;
}
h1:hover .anchor,
h2:hover .anchor,
h3:hover .anchor,
h4:hover .anchor,
h5:hover .anchor,
h6:hover .anchor {
text-decoration: none;
}
h1:hover .anchor .octicon-link,
h2:hover .anchor .octicon-link,
h3:hover .anchor .octicon-link,
h4:hover .anchor .octicon-link,
h5:hover .anchor .octicon-link,
h6:hover .anchor .octicon-link {
visibility: visible;
}
h1 {
padding-bottom: 0.3em;
font-size: 2.25em;
line-height: 1.2;
border-bottom: 1px solid #eee;
}
h1 .anchor {
line-height: 1;
}
h2 {
padding-bottom: 0.3em;
font-size: 1.75em;
line-height: 1.225;
border-bottom: 1px solid #eee;
}
h2 .anchor {
line-height: 1;
}
h3 {
font-size: 1.5em;
line-height: 1.43;
}
h3 .anchor {
line-height: 1.2;
}
h4 {
font-size: 1.25em;
}
h4 .anchor {
line-height: 1.2;
}
h5 {
font-size: 1em;
}
h5 .anchor {
line-height: 1.1;
}
h6 {
font-size: 1em;
color: #777;
}
h6 .anchor {
line-height: 1.1;
}
p,
blockquote,
ul,
ol,
dl,
table,
pre {
margin-top: 0;
margin-bottom: 16px;
}
hr {
height: 4px;
padding: 0;
margin: 16px 0;
background-color: #e7e7e7;
border: 0 none;
}
ul,
ol {
padding-left: 2em;
}
ul ul,
ul ol,
ol ol,
ol ul {
margin-top: 0;
margin-bottom: 0;
}
li>p {
margin-top: 16px;
}
dl {
padding: 0;
}
dl dt {
padding: 0;
margin-top: 16px;
font-size: 1em;
font-style: italic;
font-weight: bold;
}
dl dd {
padding: 0 16px;
margin-bottom: 16px;
}
blockquote {
padding: 0 15px;
color: #777;
border-left: 4px solid #ddd;
}
blockquote>:first-child {
margin-top: 0;
}
blockquote>:last-child {
margin-bottom: 0;
}
table {
display: block;
width: 100%;
overflow: auto;
word-break: normal;
word-break: keep-all;
}
table th {
font-weight: bold;
}
table th,
table td {
padding: 6px 13px;
border: 1px solid #ddd;
}
table tr {
background-color: #fff;
border-top: 1px solid #ccc;
}
table tr:nth-child(2n) {
background-color: #f8f8f8;
}
img {
max-width: 100%;
box-sizing: content-box;
background-color: #fff;
}
code {
padding: 0;
padding-top: 0.2em;
padding-bottom: 0.2em;
margin: 0;
font-size: 85%;
background-color: rgba(0,0,0,0.04);
border-radius: 3px;
}
code:before,
code:after {
letter-spacing: -0.2em;
content: "\00a0";
}
pre>code {
padding: 0;
margin: 0;
font-size: 100%;
word-break: normal;
white-space: pre;
background: transparent;
border: 0;
}
.highlight {
margin-bottom: 16px;
}
.highlight pre,
pre {
padding: 16px;
overflow: auto;
font-size: 85%;
line-height: 1.45;
background-color: #f7f7f7;
border-radius: 3px;
}
.highlight pre {
margin-bottom: 0;
word-break: normal;
}
pre {
word-wrap: normal;
}
pre code {
display: inline;
max-width: initial;
padding: 0;
margin: 0;
overflow: initial;
line-height: inherit;
word-wrap: normal;
background-color: transparent;
border: 0;
}
pre code:before,
pre code:after {
content: normal;
}
kbd {
display: inline-block;
padding: 3px 5px;
font-size: 11px;
line-height: 10px;
color: #555;
vertical-align: middle;
background-color: #fcfcfc;
border: solid 1px #ccc;
border-bottom-color: #bbb;
border-radius: 3px;
box-shadow: inset 0 -1px 0 #bbb;
}
.pl-c {
color: #969896;
}
.pl-c1,
.pl-s .pl-v {
color: #0086b3;
}
.pl-e,
.pl-en {
color: #795da3;
}
.pl-s .pl-s1,
.pl-smi {
color: #333;
}
.pl-ent {
color: #63a35c;
}
.pl-k {
color: #a71d5d;
}
.pl-pds,
.pl-s,
.pl-s .pl-pse .pl-s1,
.pl-sr,
.pl-sr .pl-cce,
.pl-sr .pl-sra,
.pl-sr .pl-sre {
color: #183691;
}
.pl-v {
color: #ed6a43;
}
.pl-id {
color: #b52a1d;
}
.pl-ii {
background-color: #b52a1d;
color: #f8f8f8;
}
.pl-sr .pl-cce {
color: #63a35c;
font-weight: bold;
}
.pl-ml {
color: #693a17;
}
.pl-mh,
.pl-mh .pl-en,
.pl-ms {
color: #1d3e81;
font-weight: bold;
}
.pl-mq {
color: #008080;
}
.pl-mi {
color: #333;
font-style: italic;
}
.pl-mb {
color: #333;
font-weight: bold;
}
.pl-md {
background-color: #ffecec;
color: #bd2c00;
}
.pl-mi1 {
background-color: #eaffea;
color: #55a532;
}
.pl-mdr {
color: #795da3;
font-weight: bold;
}
.pl-mo {
color: #1d3e81;
}
kbd {
display: inline-block;
padding: 3px 5px;
font: 11px Consolas, "Liberation Mono", Menlo, Courier, monospace;
line-height: 10px;
color: #555;
vertical-align: middle;
background-color: #fcfcfc;
border: solid 1px #ccc;
border-bottom-color: #bbb;
border-radius: 3px;
box-shadow: inset 0 -1px 0 #bbb;
}
.task-list-item {
list-style-type: none;
}
.task-list-item+.task-list-item {
margin-top: 3px;
}
.task-list-item input {
margin: 0 0.35em 0.25em -1.6em;
vertical-align: middle;
}
:checked+.radio-label {
z-index: 1;
position: relative;
border-color: #4078c0;
}
.sourceLine {
display: inline-block;
}
code .kw { color: #000000; }
code .dt { color: #ed6a43; }
code .dv { color: #009999; }
code .bn { color: #009999; }
code .fl { color: #009999; }
code .ch { color: #009999; }
code .st { color: #183691; }
code .co { color: #969896; }
code .ot { color: #0086b3; }
code .al { color: #a61717; }
code .fu { color: #63a35c; }
code .er { color: #a61717; background-color: #e3d2d2; }
code .wa { color: #000000; }
code .cn { color: #008080; }
code .sc { color: #008080; }
code .vs { color: #183691; }
code .ss { color: #183691; }
code .im { color: #000000; }
code .va {color: #008080; }
code .cf { color: #000000; }
code .op { color: #000000; }
code .bu { color: #000000; }
code .ex { color: #000000; }
code .pp { color: #999999; }
code .at { color: #008080; }
code .do { color: #969896; }
code .an { color: #008080; }
code .cv { color: #008080; }
code .in { color: #008080; }
</style>
<style>
body {
  box-sizing: border-box;
  min-width: 200px;
  max-width: 980px;
  margin: 0 auto;
  padding: 45px;
  padding-top: 0px;
}
</style>

</head>

<body>

<h1 id="ruegg-et-al-wifl-genoscape-data-and-code-repository">Ruegg et. al., WIFL Genoscape, Data and Code Repository</h1>
<p>Kristen Ruegg, Eric C. Anderson, Marius Somvielle, Rachael A. Bay, Christen Bossu, Mary Whitfield, Eben H. Paxton, Thomas B. Smith</p>
<ul>
<li><a href="#overview">Overview</a></li>
<li><a href="#preliminaries-and-dependencies">Preliminaries and Dependencies</a>
<ul>
<li><a href="#unix-programs">Unix Programs</a></li>
<li><a href="#genomic-resources">Genomic resources</a></li>
<li><a href="#geospatial-data">Geospatial data</a></li>
<li><a href="#java-jars">Java jars</a></li>
<li><a href="#r-packages">R Packages</a></li>
</ul></li>
<li><a href="#rmarkdown-documents">RMarkdown Documents</a>
<ul>
<li><a href="#rad-seq-data-summaries-and-pca.rmd">001-rad-seq-data-summaries-and-pca.Rmd</a></li>
<li><a href="#select-snps-for-assays-from-rad.rmd">002-select-snps-for-assays-from-rad.Rmd</a></li>
<li><a href="#literature-cited">Literature Cited</a></li>
</ul></li>
</ul>
<p><strong>Last Updated:</strong> 2020-09-22</p>
<h1 id="overview">Overview</h1>
<p>This repository includes code, data, and some intermediate results to reproduce the results in Ruegg et al. (“Seasonal niche breadth predicts population declines in a long-distance migratory bird”). You can get the whole thing by cloning or downloading the repo from <a href="https://github.com/eriqande/ruegg-et-al-wifl-genoscape">https://github.com/eriqande/ruegg-et-al-wifl-genoscape</a>.</p>
<p>If you are viewing this as the README of a GitHub repository, note that you can read it in a somewhat friendlier format (i.e., with a table of contents, etc.) on GitHub pages at: <a href="https://eriqande.github.io/ruegg-et-al-wifl-genoscape/">https://eriqande.github.io/ruegg-et-al-wifl-genoscape/</a></p>
<p>RAD sequening and the production of BAMs and VCFs was described and documented in Ruegg et al. (2018). The work described in this repository uses the products from that work</p>
<p>The various steps in our analyses are described in a series of RMarkdown documents with a fair bit of explanation/description of the procedures. The few analyses that used cluster computing resources are described and documented, but results from those analyses are included in <code>stored_results</code> so that it is not necessary to have access to a cluster to reproduce the results herein.</p>
<p>Running some of the RMarkdown documents requires a proper shell. These were initially developed and run on a Mac.</p>
<p>Rmarkdown documents and scripts in the 000 series were developed primarily by Eric C. Anderson. Rmarkdown documents and scripts in the 100 series were developed primarily by Marius Somveille.</p>
<p>All the RMarkdown documents or scripts can be evaluated en masse by sourcing the R script <code>render-numbered-Rmds.R</code>. On a fairly old mac laptop the run times for each are as follows:</p>
<pre><code>## # A tibble: 2 x 2
##   File                                    `Running time in HH:MM:SS`
##   &lt;chr&gt;                                   &lt;chr&gt;                     
## 1 001-rad-seq-data-summaries-and-pca.Rmd  00:00:45                  
## 2 002-select-snps-for-assays-from-rad.Rmd 00:00:36
</code></pre>
<h1 id="preliminaries-and-dependencies">Preliminaries and Dependencies</h1>
<p>Below is a description of the needed dependencies. A script to install all of them on one of our test clusters is at <code>000-00-prepare-dependencies.sh</code> (NOT YET INCLUDED), but it may likely need some tweaking on your system.</p>
<h2 id="unix-programs">Unix Programs</h2>
<p>The following programs must be installed and available in the PATH. Versions used on Eric’s laptop appear in parentheses. (NOT COMPLETE YET)</p>
<ul>
<li><code>bgzip</code>:</li>
<li><code>samtools</code>:</li>
<li><code>tabix</code>:</li>
</ul>
<h2 id="genomic-resources">Genomic resources</h2>
<p>Create a directory in the top level of the repository, download the willow flycatcher then modify the sequence names to be consistent with the nomenclature used in this work and index the genome (that nomenclature is already in in the FASTA file, but just not as the sequence name so, we need to only shift some words around on each sequence name line).</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1"></a><span class="fu">mkdir</span> genome</span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="bu">cd</span> genome/</span>
<span id="cb2-3"><a href="#cb2-3"></a><span class="fu">wget</span> ftp://ftp.ncbi.nlm.nih.gov/genomes/genbank/vertebrate_other/Empidonax_traillii/latest_assembly_versions/GCA_003031625.1_ASM303162v1/GCA_003031625.1_ASM303162v1_genomic.fna.gz</span>
<span id="cb2-4"><a href="#cb2-4"></a><span class="ex">ftp</span>://ftp.ncbi.nlm.nih.gov/genomes/genbank/vertebrate_other/Empidonax_traillii/latest_assembly_versions/GCA_003031625.1_ASM303162v1/GCA_003031625.1_ASM303162v1_assembly_report.txt</span>
<span id="cb2-5"><a href="#cb2-5"></a></span>
<span id="cb2-6"><a href="#cb2-6"></a><span class="co"># now, change the sequence names and make sure that the DNA bases</span></span>
<span id="cb2-7"><a href="#cb2-7"></a><span class="co"># are all listed in UPPERCASE.</span></span>
<span id="cb2-8"><a href="#cb2-8"></a><span class="ex">gzcat</span> GCA_003031625.1_ASM303162v1_genomic.fna.gz <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;</span></span>
<span id="cb2-9"><a href="#cb2-9"></a><span class="st">  /^&gt;/ {</span></span>
<span id="cb2-10"><a href="#cb2-10"></a><span class="st">    tmp = substr($1, 2)</span></span>
<span id="cb2-11"><a href="#cb2-11"></a><span class="st">    newname = $6</span></span>
<span id="cb2-12"><a href="#cb2-12"></a><span class="st">    sub(/,/, &quot;&quot;, newname)</span></span>
<span id="cb2-13"><a href="#cb2-13"></a><span class="st">    sub(/:/, &quot;|&quot;, newname)</span></span>
<span id="cb2-14"><a href="#cb2-14"></a><span class="st">    $1 = &quot;&gt;&quot; newname;</span></span>
<span id="cb2-15"><a href="#cb2-15"></a><span class="st">    $6 = tmp &quot;,&quot;;</span></span>
<span id="cb2-16"><a href="#cb2-16"></a><span class="st">    print</span></span>
<span id="cb2-17"><a href="#cb2-17"></a><span class="st">    next</span></span>
<span id="cb2-18"><a href="#cb2-18"></a><span class="st">  }</span></span>
<span id="cb2-19"><a href="#cb2-19"></a><span class="st">  {print toupper($0)}</span></span>
<span id="cb2-20"><a href="#cb2-20"></a><span class="st">&#39;</span> <span class="op">&gt;</span> wifl-genome.fna</span>
<span id="cb2-21"><a href="#cb2-21"></a><span class="ex">samtools</span> faidx wifl-genome.fna</span></code></pre></div>
<h2 id="geospatial-data">Geospatial data</h2>
<p>Make a directory in the top level of the repository called <code>geo-spatial</code> and then download some large files from Natural Earth Data to there:</p>
<ul>
<li>Natural Earth II with Shaded Relief, Water, and Drainages raster <a href="https://www.naturalearthdata.com/http//www.naturalearthdata.com/download/10m/raster/NE2_HR_LC_SR_W_DR.zip">https://www.naturalearthdata.com/http//www.naturalearthdata.com/download/10m/raster/NE2_HR_LC_SR_W_DR.zip</a>. Put or symlink the resulting directory, <code>NE2_HR_LC_SR_W_DR</code> into <code>geo-spatial</code>.</li>
<li>10m-cultural-vectors, Admin 1 – States, Provinces, Download boundary lines: <a href="https://www.naturalearthdata.com/http//www.naturalearthdata.com/download/10m/cultural/ne_10m_admin_1_states_provinces_lines.zip">https://www.naturalearthdata.com/http//www.naturalearthdata.com/download/10m/cultural/ne_10m_admin_1_states_provinces_lines.zip</a>. Put or symlink the resulting directory, <code>ne_10m_admin_1_states_provinces_lines</code> into <code>geo-spatial</code>.</li>
<li>Finally, get the coastlines: <a href="https://www.naturalearthdata.com/http//www.naturalearthdata.com/download/10m/physical/ne_10m_coastline.zip">https://www.naturalearthdata.com/http//www.naturalearthdata.com/download/10m/physical/ne_10m_coastline.zip</a>, and put the resulting folder, <code>ne_10m_coastline</code> into <code>geo-spatial</code>.</li>
</ul>
<h2 id="java-jars">Java jars</h2>
<p>The following Java-based programs must be downloaded, and the paths to their associated Jar files must be listed appropriately in the file <code>script/java-jar-paths.R</code>. (NO JAVA JARS NEEDED AT THIS TIME…)</p>
<h2 id="r-packages">R Packages</h2>
<p>Packages must be downloaded from CRAN, BioConductor, and GitHub. The following code, which is in <code>R/install_packages_etc.R</code>, will download and install the necessary packages:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="co"># just R code to install the packages needed for running the notebooks</span></span>
<span id="cb3-2"><a href="#cb3-2"></a></span>
<span id="cb3-3"><a href="#cb3-3"></a><span class="co"># get the packages needed from CRAN</span></span>
<span id="cb3-4"><a href="#cb3-4"></a><span class="kw">install.packages</span>(</span>
<span id="cb3-5"><a href="#cb3-5"></a>  <span class="kw">c</span>(</span>
<span id="cb3-6"><a href="#cb3-6"></a>    <span class="st">&quot;hms&quot;</span>,</span>
<span id="cb3-7"><a href="#cb3-7"></a>    <span class="st">&quot;knitr&quot;</span>,</span>
<span id="cb3-8"><a href="#cb3-8"></a>    <span class="st">&quot;lubridate&quot;</span>,</span>
<span id="cb3-9"><a href="#cb3-9"></a>    <span class="st">&quot;rmarkdown&quot;</span>,</span>
<span id="cb3-10"><a href="#cb3-10"></a>    <span class="st">&quot;rubias&quot;</span>,</span>
<span id="cb3-11"><a href="#cb3-11"></a>    <span class="st">&quot;sessioninfo&quot;</span>,</span>
<span id="cb3-12"><a href="#cb3-12"></a>    <span class="st">&quot;sf&quot;</span>,</span>
<span id="cb3-13"><a href="#cb3-13"></a>    <span class="st">&quot;tidyverse&quot;</span></span>
<span id="cb3-14"><a href="#cb3-14"></a>  ),</span>
<span id="cb3-15"><a href="#cb3-15"></a>  <span class="dt">repos =</span> <span class="st">&quot;http://cran.rstudio.com&quot;</span></span>
<span id="cb3-16"><a href="#cb3-16"></a>)</span>
<span id="cb3-17"><a href="#cb3-17"></a></span>
<span id="cb3-18"><a href="#cb3-18"></a><span class="co"># Install packages from BioConductor</span></span>
<span id="cb3-19"><a href="#cb3-19"></a><span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>))</span>
<span id="cb3-20"><a href="#cb3-20"></a>  <span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</span>
<span id="cb3-21"><a href="#cb3-21"></a></span>
<span id="cb3-22"><a href="#cb3-22"></a>BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;SNPRelate&quot;</span>)</span>
<span id="cb3-23"><a href="#cb3-23"></a></span>
<span id="cb3-24"><a href="#cb3-24"></a></span>
<span id="cb3-25"><a href="#cb3-25"></a></span>
<span id="cb3-26"><a href="#cb3-26"></a><span class="co"># Install Eric&#39;s Packages from GitHub</span></span>
<span id="cb3-27"><a href="#cb3-27"></a><span class="co"># UPDATE UPON FINAL RELEASE WITH THE ACTUAL COMMIT USED</span></span>
<span id="cb3-28"><a href="#cb3-28"></a>remotes<span class="op">::</span><span class="kw">install_github</span>(<span class="st">&quot;eriqande/genoscapeRtools&quot;</span>)</span>
<span id="cb3-29"><a href="#cb3-29"></a>remotes<span class="op">::</span><span class="kw">install_github</span>(<span class="st">&quot;eriqande/snps2assays&quot;</span>)</span>
<span id="cb3-30"><a href="#cb3-30"></a>remotes<span class="op">::</span><span class="kw">install_github</span>(<span class="st">&quot;eriqande/whoa&quot;</span>)</span></code></pre></div>
<h1 id="rmarkdown-documents">RMarkdown Documents</h1>
<p>The following RMarkdown documents should be evaluated in order. The script <code>render-numbered-Rmds.R</code> (NOT WRITTEN YET) will do that when run in the top level of this repository. Some RMarkdown documents rely on outputs from previous ones. Some of these RMarkdown documents include calls to Unix utilities, so might not run on non-Unix or non-Linux architectures.</p>
<p>Outputs (figures, tables, R data objects, etc) from each RMarkdown document are written to the <code>outputs/XXX</code> directories. Intermediate files written during evaluation of each RMarkdown document are written to the <code>intermediates/XXX</code> directories (THOUGH I MIGHT NOT USE intermediates AT ALL). To facilitate working between the cluster and a desktop/laptop, some outputs are written to the <code>stored_results/XXX</code> directories which are version controlled and included in this repo.</p>
<p>Thumbnails of the figures and tables generated by each RMarkdown document appear below (NOT YET). Additionally, at the top of each section is a link to the compiled (HTML-version) of the RMarkdown document on GitHub Pages.</p>
<h2 id="001-rad-seq-data-summaries-and-pcarmd">001-rad-seq-data-summaries-and-pca.Rmd</h2>
<p>(<em>Compiled RMarkdown HTML document on GitHub Pages:</em> <a href="https://eriqande.github.io/ruegg-et-al-wifl-genoscape/001-rad-seq-data-summaries-and-pca.html">001-rad-seq-data-summaries-and-pca.html</a>)</p>
<p>Summarizing the RAD seq data, filtering out obvious paralogs and monomorphic sites, then producing a PCA plot and estimates of pairwise (F_\mathrm{ST}).</p>
<h2 id="002-select-snps-for-assays-from-radrmd">002-select-snps-for-assays-from-rad.Rmd</h2>
<p>(<em>Compiled RMarkdown HTML document on GitHub Pages:</em> <a href="https://eriqande.github.io/ruegg-et-al-wifl-genoscape/002-select-snps-for-assays-from-rad.html">002-select-snps-for-assays-from-rad.html</a>)</p>
<p>All the steps we went through to rank SNPs for ability to resolve different populations/subspecies, and then the workflow for designing Fluidigm assays from them.</p>
<h2 id="literature-cited">Literature Cited</h2>
<div id="refs" class="references">

<div id="ref-ruegg2018ecological">

<p>Ruegg, Kristen, Rachael A Bay, Eric C Anderson, James F Saracco, Ryan J Harrigan, Mary Whitfield, Eben H Paxton, and Thomas B Smith. 2018. “Ecological Genomics Predicts Climate Vulnerability in an Endangered Southwestern Songbird.” <em>Ecology Letters</em> 21 (7): 1085–96.</p>
</div>

</div>

</body>
</html>
